### PR TITLE
List engines API

### PR DESCRIPTION
**Pull request by carlosdelest**
Implement List Engines API:

`GET /_engine`

```json
{
  "_meta": {
    "from": 0,
    "size": 100,
    "total": 3
  },
  "results": [
    {
      "name": "test-engine",
      "indices": [
        "test1",
        "test2"
      ],
      "engine_alias": "engine-test-engine",
      "updated_at_millis": 1677524160042
    },
    {
      "name": "test-engine-another",
      "indices": [
        "test2"
      ],
      "engine_alias": "engine-test-engine-another",
      "updated_at_millis": 1677524160042
    },
    {
      "name": "test-engine-another-one",
      "indices": [
        "test1"
      ],
      "analytics_collection_name": "analytics-collection-1",
      "engine_alias": "engine-test-engine-another-one",
      "updated_at_millis": 1677524160042
    }
  ]
}
```

The following query string parameters are available:

- Pagination: `from` and `size`
`GET /_engine?from=1&size=1`

```json
{
  "_meta": {
    "from": 1,
    "size": 1,
    "total": 3
  },
  "results": [
    {
      "name": "test-engine-another",
      "indices": [
        "test2"
      ],
      "engine_alias": "engine-test-engine-another",
      "updated_at_millis": 1677524141942
    }
  ]
}
```

- Query:

`GET /_engine/?q=*another*`

```json
{
  "_meta": {
    "from": 0,
    "size": 100,
    "total": 2
  },
  "results": [
    {
      "name": "test-engine-another",
      "indices": [
        "test2"
      ],
      "engine_alias": "engine-test-engine-another",
      "updated_at_millis": 1677524186957
    },
    {
      "name": "test-engine-another-one",
      "indices": [
        "test1"
      ],
      "analytics_collection_name": "analytics-collection-1",
      "engine_alias": "engine-test-engine-another-one",
      "updated_at_millis": 1677524186957
    }
  ]
}
```

Pending:
- [ ] Serialization tests
- [ ] REST tests

### CODE REVIEW

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java
Comment (demjened) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:344]: Agreed - I like the abstraction here (`listEngines()` returns a list of `Engine`s, regardless of where they came from).

Also, is it a convention to use `Tuple`s for compound return types, rather than introducing a new class?

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexServiceTests.java
Comment (demjened) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexServiceTests.java:117]: I wonder if this notation is the preferred Elasticsearch convention, rather than using the shorter `assertEquals(10, engines.length)`.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (demjened) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:1]: I don't see any unit tests for classes in the `action` package, I'm guessing we don't do that in the ES codebase. Are there any other means of testing their functionality?

**Review by demjened — APPROVED** [review]
LGTM with some questions.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:344]: I agree with @demjened , a private class might feel a bit cleaner here.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:1]: @demjened we've been doing REST yaml tests and serialization tests. I believe @carlosdelest will be adding these shortly, but they're pretty boilerplate. If you would like to look at an example you can see them in PRs such as [this one](https://github.com/elastic/elasticsearch/pull/94090)

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:147]: I think we could omit this overridden method?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:29]: Nitpick: We called this the List Engines call in the Rails API. However, would it make sense to call this `SearchEnginesAction` instead, since we're effectively searching every call?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:65]: Do we want to do any validation on the query, for example non-empty, or non-allowed characters in an Engine name?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:84]: What is the impact of forbidding closed here? Does this mean this call will fail if underlying engine indices are closed?

**Review by kderusso — COMMENTED** [review]
Overall looks good! I left a few comments, and look forward to reviewing more after all tests are implemented. Thank you!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:344]: Sounds good! I will implement that as a static private class. Thanks! 👍

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:29]: I think List describes the action accurately, but I'm open to a name change.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:65]: I punted on this as I didn't find the query string validation mechanism in ES Search. I'll check again 👍

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:84]: It's the Engine system index the one that will be the one queried, so I was thinking on having the strictest set of IndexOptions I could find. The System index is not expected to be closed, so I guess this is good.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:147]: It's part of the `StatusToXContentObject` interface, it needs to be implemented.

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexServiceTests.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexServiceTests.java:117]: good call! I guess this is more readable in terms of expected / actual value. Maybe Jim has some comments here on styling.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:32]: I don't have really strong feelings about it, it's OK to keep as list.

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineTestUtils.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineTestUtils.java:18]: Nice 👍

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineTestUtils.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineTestUtils.java:20]: Nice 👍

**Review by kderusso — COMMENTED** [review]
Looking good! I think we just need some yaml tests and this will be good to go!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:111]: This object is used for requests and I don't think it should be exposed in the response. There's a `QueryPage` object that I think would fit better to expose the total number of result and the result itself. We don't need to repeat the size and from in the response imo.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:313]: I am not sure we should use the `Engine` class in the response of this API. We won't return the entire `Engine` object, it's possible currently because we only have a name, a list of indices and an analytics collection name but in the near future we'll have more configurations and they won't be returned in the list API.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:42]: That should be good as long as we use the right index options (that disables index pattern expansion).

**Review by jimczi — COMMENTED** [review]
I left some comments, thanks @carlosdelest

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:155]: That won't work if we don't add this field in the rest parser too (see the failing tests in the PR). Although I am not sure we want to do that since we'd need to ensure that the name matches the id that is provided in the request. Can we remove it for now and discuss if this is really needed in a follow up?

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexServiceTests.java
Comment (jimczi) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexServiceTests.java:116]: I also find the hamcrest version more readable too, it also gives better error message. I don't think we have a strict convention here though.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:86]: @carlosdelest I think this needs to be the engine alias and that's why the yaml tests are failing.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:313]: Right now the response returns the entire Engine. If the EngineIndexService does not retrieve the Engines, then the Transport action will need to do so, right?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:86]: This is pointing already to the engine index alias name. Do you mean to the individual engines indices?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:111]: That's what the current API is doing, although I guess the `size` and `from` are not needed. I'll check the QueryPage object, thanks!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:155]: We could remove that, but then the API would need to be modified, as there would be no ID returned for each of the Engines returned (and thus we wouldn't be able to identify them).

I'll check adding them to the parser, happy to hear alternatives / better response formats, even though it might impact the frontend work.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:313]: I don't think we should return the entire Engine here. The list API can be used to list the engine names, their associated indices and maybe their analytics collections names. To get the entire Engine configuration, another call to get the Engine by name should be needed (using the get API).

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:86]: This won't work since we don't have any index name to pass here (it's a list API). I guess that means that we need to add a new cluster privilege to manage Engine? @ioanatia what's your take on this? We assumed that using the index privilege was enough for now but that won't work for this API. Maybe we can start with reusing the cluster monitor privilege for this API and continue the discussion async in order to not block this PR?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:155]: I think it would be more complete to add it to the response.

Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:155]: Ok so let's add it as an optional param in the parser and we can validate that it matches the expected name when we build the Engine object? We should also do the same for the alias name since it's accepted by the parser currently but we don't check if the alias name is correct.

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_engine_put.yml
Comment (carlosdelest) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/20_engine_put.yml:39]: Added teardown to tests, as they may impact the list engines testing

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/30_engine_get.yml
Comment (carlosdelest) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/30_engine_get.yml:40]: Name is added to returned results now

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:97]: Check the engine name in case it is present, so it matches the resource name

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:168]: Added in the parser, please check the solution 👍

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:1]: Added REST and serialization tests, please check 👍

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:67]: Added query validation - non empty

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (carlosdelest) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:105]: Used QueryPage in the response per Jim's suggestion. It makes sense to use other Elasticsearch constructs, and making responses in line with other ES APIs instead of adding `_meta` like Ent Search did

File: rest-api-spec/src/main/resources/rest-api-spec/api/engine.list.json
Comment (kderusso) [inline rest-api-spec/src/main/resources/rest-api-spec/api/engine.list.json:34]: Should `from` be an `int` and not a `number`?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java
Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java:302]: Can we create a special class for this like `EngineListResponse`? As I said in my previous comment I'd like to avoid returning the entire Engine. It might be ok now but we'll add more configurations later and we shouldn't return the full object in this API. Let's restrict the response to the indices names, the alias name and the analytics collection name? I know that it's the entire Engine configuration now but this will change in the near future.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:40]: This can be removed now

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/50_engine_list.yml
Comment (kderusso) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/50_engine_list.yml:59]: The Rails API limited engines to those we had permissions to test against. So if there existed a `test-engine-3` that we didn't have permissions on it wouldn't be returned in results. 

I don't _think_ we have to explicitly test that here since the index permissions should take care of it for us? But I wanted to ask just in case.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:79]: There's a validation that fails if the query is null so I guess we can use `writeString` directly?

**Review by jimczi — COMMENTED** [review]
It looks great, thanks @carlosdelest . I left one comment regarding the extent of the response that I'd like to see fixed before merging.

**Review by kderusso — APPROVED** [review]
LGTM, nice work!

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/50_engine_list.yml
Comment (carlosdelest) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/50_engine_list.yml:68]: That's because I used the JSON splat operator on... No, just kidding, it's a copypasta error 😅 . Thanks for noticing!

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/Engine.java:97]: engineName can be null, for example on a PUT where the body that does not include the engine name. It is a valid situation.

File: rest-api-spec/src/main/resources/rest-api-spec/api/engine.list.json
Comment (carlosdelest) [reply rest-api-spec/src/main/resources/rest-api-spec/api/engine.list.json:34]: Right! Thanks for spotting this 👍

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexServiceTests.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexServiceTests.java:127]: That case is not possible I think - query is either default value or some non-null, non-empty query issued by the user

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/50_engine_list.yml
Comment (carlosdelest) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/50_engine_list.yml:59]: Someone with cluster_manage_engines will be able to see all engines (list and get). When they want to update them, then index permissions will kick in.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:69]: As this will be validated by the actual search query, I'll punt on this - I haven't found a validation for query strings in the codebase. Happy to do that in a follow-up PR if anyone point me to that!

Comment (jimczi) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:69]: That's fine, the validation is done when the query is built on a per-shard basis so we can rely on search failure to be raised if the syntax is not correct.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java
Comment (carlosdelest) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/EngineIndexService.java]: Done! LMKWYT

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java
Comment (jimczi) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/ListEnginesAction.java:48]: nit: `requireNonNullElse` is not needed since readString cannot return null.

**Review by jimczi — APPROVED** [review]
Thanks for all the iterations @carlosdelest , LGTM!

**Pull request merged**
